File: Sources/Sentry/SentryTransportAdapter.m
#import "SentryTransportAdapter.h"
#import "SentryEnvelope.h"
#import "SentryEvent.h"
#import "SentryOptions.h"
#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@interface
SentryTransportAdapter ()

@property (nonatomic, strong) id<SentryTransport> transport;
@property (nonatomic, strong) SentryOptions *options;

@end

@implementation SentryTransportAdapter

- (instancetype)initWithTransport:(id<SentryTransport>)transport options:(SentryOptions *)options
{
    if (self = [super init]) {
        self.transport = transport;
        self.options = options;
    }

    return self;
}

- (void)sendEvent:(SentryEvent *)event attachments:(NSArray<SentryAttachment *> *)attachments
{
    [self sendEvent:event traceState:nil attachments:attachments];
}

- (void)sendEvent:(SentryEvent *)event
          session:(SentrySession *)session
      attachments:(NSArray<SentryAttachment *> *)attachments
{
    [self sendEvent:event withSession:session traceState:nil attachments:attachments];
}

- (void)sendEvent:(SentryEvent *)event
       traceState:(nullable SentryTraceState *)traceState
      attachments:(NSArray<SentryAttachment *> *)attachments
{
    [self sendEvent:event
                     traceState:traceState
                    attachments:attachments
        additionalEnvelopeItems:@[]];
}

- (void)sendEvent:(SentryEvent *)event
                 traceState:(nullable SentryTraceState *)traceState
                attachments:(NSArray<SentryAttachment *> *)attachments
    additionalEnvelopeItems:(NSArray<SentryEnvelopeItem *> *)additionalEnvelopeItems
{
    NSMutableArray<SentryEnvelopeItem *> *items = [self buildEnvelopeItems:event
                                                               attachments:attachments];
    [items addObjectsFromArray:additionalEnvelopeItems];

    SentryEnvelopeHeader *envelopeHeader = [[SentryEnvelopeHeader alloc] initWithId:event.eventId
                                                                         traceState:traceState];
    SentryEnvelope *envelope = [[SentryEnvelope alloc] initWithHeader:envelopeHeader items:items];

    [self sendEnvelope:envelope];
}

- (void)sendEvent:(SentryEvent *)event
      withSession:(SentrySession *)session
       traceState:(nullable SentryTraceState *)traceState
      attachments:(NSArray<SentryAttachment *> *)attachments
{
    NSMutableArray<SentryEnvelopeItem *> *items = [self buildEnvelopeItems:event
                                                               attachments:attachments];
    [items addObject:[[SentryEnvelopeItem alloc] initWithSession:session]];

    SentryEnvelopeHeader *envelopeHeader = [[SentryEnvelopeHeader alloc] initWithId:event.eventId
                                                                         traceState:traceState];

    SentryEnvelope *envelope = [[SentryEnvelope alloc] initWithHeader:envelopeHeader items:items];

    [self sendEnvelope:envelope];
}

- (void)sendUserFeedback:(SentryUserFeedback *)userFeedback
{
    SentryEnvelope *envelope = [[SentryEnvelope alloc] initWithUserFeedback:userFeedback];
    [self sendEnvelope:envelope];
}

- (void)sendEnvelope:(SentryEnvelope *)envelope
{
    [self.transport sendEnvelope:envelope];
}

- (void)recordLostEvent:(SentryDataCategory)category reason:(SentryDiscardReason)reason
{
    [self.transport recordLostEvent:category reason:reason];
}

- (NSMutableArray<SentryEnvelopeItem *> *)buildEnvelopeItems:(SentryEvent *)event
                                                 attachments:
                                                     (NSArray<SentryAttachment *> *)attachments
{
    NSMutableArray<SentryEnvelopeItem *> *items = [NSMutableArray new];
    [items addObject:[[SentryEnvelopeItem alloc] initWithEvent:event]];

    for (SentryAttachment *attachment in attachments) {
        SentryEnvelopeItem *item =
            [[SentryEnvelopeItem alloc] initWithAttachment:attachment
                                         maxAttachmentSize:self.options.maxAttachmentSize];
        // The item is nil, when creating the envelopeItem failed.
        if (nil != item) {
            [items addObject:item];
        }
    }

    return items;
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Networking/SentryTransportAdapter.swift
// swiftlint:disable missing_docs
import Foundation

@_spi(Private)
@objc(SentryTransportAdapter)
open class SentryTransportAdapter: NSObject {

    private let transports: [any Transport]
    private let options: Options

    @objc(initWithTransports:options:)
    public init(transports: [any Transport], options: Options) {
        self.transports = transports
        self.options = options
        super.init()
    }

    @objc(sendEvent:session:attachments:)
    public func send(_ event: Event, session: SentrySession, attachments: [Attachment]) {
        send(event, with: session, traceContext: nil, attachments: attachments)
    }

    @objc(sendEvent:traceContext:attachments:)
    open func send(event: Event, traceContext: TraceContext?, attachments: [Attachment]) {
        send(event: event, traceContext: traceContext, attachments: attachments, additionalEnvelopeItems: [])
    }

    @objc(sendEvent:traceContext:attachments:additionalEnvelopeItems:)
    open func send(
        event: Event,
        traceContext: TraceContext?,
        attachments: [Attachment],
        additionalEnvelopeItems: [SentryEnvelopeItem]
    ) {
        var items = buildEnvelopeItems(event, attachments: attachments)
        items.append(contentsOf: additionalEnvelopeItems)

        let envelope = SentryEnvelope(
            header: SentryEnvelopeHeader(id: event.eventId, traceContext: traceContext),
            items: items
        )
        send(envelope: envelope)
    }

    @objc(sendEvent:withSession:traceContext:attachments:)
    open func send(
        _ event: Event,
        with session: SentrySession,
        traceContext: TraceContext?,
        attachments: [Attachment]
    ) {
        var items = buildEnvelopeItems(event, attachments: attachments)
        items.append(SentryEnvelopeItem(session: session))

        let envelope = SentryEnvelope(
            header: SentryEnvelopeHeader(id: event.eventId, traceContext: traceContext),
            items: items
        )
        send(envelope: envelope)
    }

    @objc(storeEvent:traceContext:)
    open func store(_ event: Event, traceContext: TraceContext?) {
        let envelope = SentryEnvelope(
            header: SentryEnvelopeHeader(id: event.eventId, traceContext: traceContext),
            items: [SentryEnvelopeItem(event: event)]
        )
        transports.forEach { $0.store(envelope) }
    }

    @objc(sendEnvelope:)
    public func send(envelope: SentryEnvelope) {
        transports.forEach { $0.send(envelope: envelope) }
    }

    @objc(recordLostEvent:reason:)
    public func recordLostEvent(_ category: SentryDataCategory, reason: SentryDiscardReason) {
        transports.forEach { $0.recordLostEvent(category, reason: reason) }
    }

    @objc(recordLostEvent:reason:quantity:)
    public func recordLostEvent(_ category: SentryDataCategory, reason: SentryDiscardReason, quantity: UInt) {
        transports.forEach { $0.recordLostEvent(category, reason: reason, quantity: quantity) }
    }

    @objc(flush:)
    public func flush(_ timeout: TimeInterval) {
        transports.forEach { _ = $0.flush(timeout) }
    }

    private func buildEnvelopeItems(_ event: Event, attachments: [Attachment]) -> [SentryEnvelopeItem] {
        [SentryEnvelopeItem(event: event)] + attachments.compactMap {
            SentryEnvelopeItem(attachment: $0, maxAttachmentSize: options.maxAttachmentSize)
        }
    }
}
// swiftlint:enable missing_docs
